### ISSUE TITLE
Instructions for connecting to gateway outside of helm

### ISSUE THREAD

**Issue by chrisroat** [dask/dask-gateway#337]
I am using the daskhub chart now, and I also have some non-jupyter workers that do dask work.  I am trying to have them connect to the gateway, but not quite having success.

Should I be updating the daskhub chart to use the SimpleAuthenticator (even though it's not recommended)?

Or can I still use jupyterhub auth, even though my worker isn't logged into jhub?  If I set JUPYTERHUB_API_TOKEN, DASK_GATEWAY__ADDRESS, and DASK_GATEWAY__AUTH__TYPE, I don't get rejected, but I end up crashing the gateway server because of a KeyError on the "groups" field that is referenced in this line:

https://github.com/dask/dask-gateway/blob/1d00e87ae57aae88b08f321e538922dcdc6e004e/dask-gateway-server/dask_gateway_server/auth.py#L393

**Comment by chrisroat** [comment]
I forgot to also mention that I tried the following and using the LoadBalancer ip as my gateway address, and I just get a 404.

```
dask-gateway:
  traefik:
    service:
      type: LoadBalancer
```

**Comment by jcrist** [comment]
> Should I be updating the daskhub chart to use the SimpleAuthenticator (even though it's not recommended)?

You should keep the jupyterhub authenticator, but have the users [create an api token](https://jupyterhub.readthedocs.io/en/stable/reference/rest.html#create-an-api-token). You can then configure the gateway client to use that token. You have a few options here:

- Set the token as `JUPYTERHUB_API_TOKEN` in the environment
- Manually pass the token to `JupyterHubAuth` when creating a client
  ```python
  from dask_gateway import JupyterHubAuth, Gateway
  gateway = Gateway(auth=JupyterHubAuth(api_token=...)
  ```

- Configure the token statically in your dask config
  ```yaml
  gateway:
    auth:
      type: jupyterhub
      kwargs:
        api_token: <your token>
  ```

> If I set JUPYTERHUB_API_TOKEN, DASK_GATEWAY__ADDRESS, and DASK_GATEWAY__AUTH__TYPE, I don't get rejected, but I end up crashing the gateway server because of a KeyError on the "groups" field

Hmm, that's interesting, and might be a bug (in our code, or in the jupyterhub code. I expected `groups` to always be present in the returned user model. What version of JupyterHub?

> I forgot to also mention that I tried the following and using the LoadBalancer ip as my gateway address, and I just get a 404.

That error looks unrelated (traefik complains if the service isn't ready, which happens far too often for something it handles fine). Are you connecting through the jupyterhub proxy (`you -> jhub proxy -> traefik -> dask-gateway`), or directly through the exposed traefik proxy? Do you still have the `/services/dask-gateway` prefix in your url? You get a 404 after doing what?

**Comment by chrisroat** [comment]
> You should keep the jupyterhub authenticator, but have the users [create an api token](https://jupyterhub.readthedocs.io/en/stable/reference/rest.html#create-an-api-token). You can then configure the gateway client to use that token. You have a few options here:
> 
> * Set the token as `JUPYTERHUB_API_TOKEN` in the environment

Yes, this is exactly what I was trying.  You mention "users" (plural), but I believe there is only a single token that anyone can use, correct?

This is the env yaml for my worker, where I put the token in a secret.

```
            - name: JUPYTERHUB_API_TOKEN
              valueFrom:
                secretKeyRef:
                  name: jhub-secrets
                  key: JHUB_TOKEN
            - name: DASK_GATEWAY__ADDRESS
              value: http://proxy-http:8000/services/dask-gateway
            - name: DASK_GATEWAY__AUTH__TYPE
              value: jupyterhub
```

I am using jupyterhub 1.1.0 and jupyterlab 2.2.8.

> You get a 404 after doing what?

Trying to use the LoadBalancer IP.  I did not put the `/services/dask-gateway` prefix, and that was my mistake.  I can do that.  Should I expect to have more success going through the LoadBalancer IP than the proxy-http:8000?  (My worker is in the K8s cluster, so I don't technically need the external IP)

**Comment by jcrist** [comment]
> You mention "users" (plural), but I believe there is only a single token that anyone can use, correct?

Ah, that might be why you're getting the `groups` error - here you want an api token associated with a *user's* account, not the token associated with the *dask-gateway server*. Every user in a JupyterHub instance can have 0 or more api tokens associated with them, these can be used to login programmatically rather than going through the login page. Can you try going through the jupyterhub UI to create an api token and use that to authenticate?

> Should I expect to have more success going through the LoadBalancer IP than the proxy-http:8000?

It looks like the thing you're trying to connect to the dask-gateway server is also running inside the same k8s cluster? If so, you don't need a `LoadBalancer` for dask-gateway at all, you can use the same addresses as you would from within a jupyterhub pod. It's only if you're connecting from outside the k8s cluster that you'd need a load balancer on traefik (and in that configuration it'd be preferable to go directly through traefik rather than hopping through jupyterhub as well).

**Comment by chrisroat** [comment]
So many tokens, so little time!

Thanks, that got things working!

**Comment by chrisroat** [comment]
BTW, I wanted to bring up my workers and jhub at the same time, so wanted to pass both a pregenerated token.  I tried to use services for this, but could not get it to work because dask-gateway looks for groups, and services don't have groups.

To make things work, I had to whitelist my new user in my yaml config, and then create a group and an api_token (which are soft deprecated):

          c.JupyterHub.load_groups.update({'worker': ['worker']})
          c.JupyterHub.api_tokens = {os.environ['WORKER_TOKEN']: 'worker'}

Could dask-gateway accept a user or service without groups?
